crates/re_viewer_context/src/store_hub.rs

```diff
@@ -467,48 +530,36 @@ impl StoreHub {
     /// Try to load the persisted blueprint for the given `ApplicationId`.
     /// Note: If no blueprint exists at the expected path, the result is still considered `Ok`.
     /// It is only an `Error` if a blueprint exists but fails to load.
-    // TODO(#2579): implement persistence for web
-    #[cfg(not(target_arch = "wasm32"))]
-    pub fn try_to_load_persisted_blueprint(
-        &mut self,
-        app_id: &ApplicationId,
-    ) -> anyhow::Result<()> {
-        use crate::blueprint::is_valid_blueprint;
-
+    fn try_to_load_persisted_blueprint(&mut self, app_id: &ApplicationId) -> anyhow::Result<()> {
         re_tracing::profile_function!();
-        let blueprint_path = default_blueprint_path(app_id)?;
-        if blueprint_path.exists() {
-            re_log::debug!("Trying to load blueprint for {app_id} from {blueprint_path:?}",);
-
-            let with_notifications = false;
-            if let Some(mut bundle) = load_blueprint_file(&blueprint_path, with_notifications) {
-                for store in bundle.drain_entity_dbs() {
-                    if store.store_kind() == StoreKind::Blueprint && store.app_id() == Some(app_id)
-                    {
-                        if !is_valid_blueprint(&store) {
-                            re_log::warn_once!("Blueprint for {app_id} appears invalid - restoring to default. This is expected if you have just upgraded Rerun versions.");
-                            continue;
-                        }
-                        // We found the blueprint we were looking for; make it active.
-                        // borrow-checker won't let us just call `self.set_blueprint_for_app_id`
-                        re_log::debug!(
-                            "Switching blueprint for {app_id} to {} loaded from {blueprint_path:?}",
-                            store.store_id(),
-                        );
-                        self.active_blueprint_by_app_id
-                            .insert(app_id.clone(), store.store_id().clone());
-                        self.blueprint_last_save
-                            .insert(store.store_id().clone(), store.generation());
-                        self.store_bundle.insert(store);
-                    } else {
-                        anyhow::bail!(
-                            "Found unexpected store while loading blueprint: {:?}",
-                            store.store_id()
-                        );
-                    }
+
+        let Some(loader) = &self.persistence.loader else {
+            return Ok(());
+        };
+
+        if let Some(mut bundle) = (loader)(app_id)? {
+            for store in bundle.drain_entity_dbs() {
```

Comment:
Idea (not for this PR). It could be nice if this bundle actually contained **both** the active and the blueprint it was cloned from. We should be able to write both of them to the store on save.

We probably need to modify the bundle-reader to look for and track the BlueprintActivationCommand messages.

Reply:
That's a pretty interesting idea